### PR TITLE
Add conformance automation and manifest publishing

### PR DESCRIPTION
**Pull request by phoracek**
**What this PR does / why we need it**:

With this patch, sonobuoy conformance manifest will be published as part of released artifacts.

This patch also introduce a new makefile target to execute conformance tests against the development cluster.

Finally, this patch adds a new automation script to exercise conformance suite on CI.

Nightly prow job exercising this suite: https://github.com/kubevirt/project-infra/pull/623

**Special notes for your reviewer**:

/hold

This patch updates the builder image. We need to publish a new version of it to our registry before merging this PR.

**Release note**:

```release-note
Add conformance automation and manifest publishing
```

### CODE REVIEW

File: hack/builder/Dockerfile [hunk 1]
```diff
@@ -92,6 +92,12 @@ RUN set -x && \
 
 RUN pip3 install j2cli operator-courier==1.3.0
 
+RUN set -x && \
+    wget https://github.com/vmware-tanzu/sonobuoy/releases/download/v0.19.0/sonobuoy_0.19.0_linux_amd64.tar.gz && \
+    tar xvf sonobuoy_0.19.0_linux_amd64.tar.gz && \
```
Comment (oshoval) [inline hack/builder/Dockerfile:97]: on kubevirtci, it always take the latest version of sonobuoy isnt it ?
i think we should do the same and not have a fixed version ?

File: hack/conformance.sh [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+echo 'Results:'
+sonobuoy results ${results_archive}
+
+echo "Extracting the full report and keep it under ${ARTIFACTS}"
+tar xf ${results_archive} -C ${ARTIFACTS}
+cp ${ARTIFACTS}/plugins/kubevirt-conformance/results/global/junit.xml ${ARTIFACTS}/junit.conformance.xml
+
+echo 'Evaluating success of the test run'
+sonobuoy results ${results_archive} | grep -q 'Status: passed' || { echo 'Conformance suite has failed'; exit 1; }
```
Comment (oshoval) [inline hack/conformance.sh:30]: there are at least 3 fields of Status, all should pass, what if one passed and two failed ?
please see what we have done on kubevirtci 
```
    passed=$(./sonobuoy status --json | jq  ' .plugins[] | select(."result-status" == "passed")'  | wc -l)
    failed=$(./sonobuoy status --json | jq  ' .plugins[] | select(."result-status" == "failed")'  | wc -l)

    if [ $passed -eq 0 ] || [ $failed -ne 0 ]; then
        echo "sonobuoy failed"
        exit 1
    fi
```

File: automation/conformance.sh [hunk 3]
```diff
@@ -0,0 +1,13 @@
+#!/usr/bin/env bash
+
+set -euo pipefail
+
+export WORKSPACE="${WORKSPACE:-$PWD}"
+readonly ARTIFACTS_PATH="${ARTIFACTS-$WORKSPACE/exported-artifacts}"
+mkdir -p "$ARTIFACTS_PATH"
+
+trap "{ make cluster-down; cp -r _out/artifacts/conformance ${ARTIFACTS_PATH}; }" EXIT SIGINT SIGTERM
```
Comment (oshoval) [inline automation/conformance.sh:9]: SIGTERM isnt needed because we have EXIT, right ?

**Review by oshoval — CHANGES_REQUESTED** [review]

File: automation/conformance.sh (hunk 3 above)
Comment (phoracek) [reply automation/conformance.sh:9]: I don't know. Any source?

File: hack/builder/Dockerfile (hunk 1 above)
Comment (phoracek) [reply hack/builder/Dockerfile:97]: All the other dependencies here are pinned to a specific version. The builder image is built and published manually, so I assume we want to have close control over our dependencies here.

Also, we don't run Kubernetes conformance tests. We only use sonobuoy to run our own conformance plugin.

File: hack/conformance.sh (hunk 2 above)
Comment (phoracek) [reply hack/conformance.sh:30]: I see only one:

```
Plugin: kubevirt-conformance
Status: passed
Total: 580
Passed: 1
Failed: 0
Skipped: 579
```

I'm retrieving the information from downloaded results archive, while your command references to status.

File: automation/conformance.sh (hunk 3 above)
Comment (oshoval) [reply automation/conformance.sh:9]: Quique raised it once, 
according this unformal link, it depends on the running shell,
https://unix.stackexchange.com/questions/57940/trap-int-term-exit-really-necessary
you can either try it, leave it as it, or we can find an official source. not crucial.

Comment (phoracek) [reply automation/conformance.sh:9]: One of the comments:

```
dash also doesn't trap on just EXIT when it receives SIGINT/SIGTERM
```

We use dash on Prow.

Comment (oshoval) [reply automation/conformance.sh:9]: you rest your case ;]

Comment (oshoval) [reply automation/conformance.sh:9]: you rest your case ;]
We will have this problem on more scripts, where SIGTERM was omitted

Comment (oshoval) [reply automation/conformance.sh:9]: btw do you want to add SIGQUIT ? 
same as SIGINT but with core dump

Comment (oshoval) [reply automation/conformance.sh:9]: one note 
in case `make cluster-down` fails, its error code would mask the real error code in case the script failed
in case its important for you

Comment (phoracek) [reply automation/conformance.sh:9]: I don't care which RC will be returned as long as it is a failure.

I don't think we need to to work with the expectation of code dump. Do you see a realistic scenario where core dump would happen and it would make sense (or be possible) to execute the trap?

Comment (oshoval) [reply automation/conformance.sh:9]: just if user triggered 

SIGQUIT is the dump core signal. The terminal sends it to the foreground process when the user presses ctrl-\. The default behavior is to terminate the process and dump core, but it can be caught or ignored. The intention is to provide a mechanism for the user to abort the process. You can look at SIGINT as "user-initiated happy termination" and SIGQUIT as "user-initiated unhappy termination."

I think if we would have a common agreed trap line it would save us time, but up to you.

Btw see PR 428, where there was a shebang that enforce it to run with bash if you want, and then you might can remove the SIGTERM, but its up to you.
(not sure if every node has bash and so on, in case it doesnt, then it wont be good to add it of course)

Comment (phoracek) [reply automation/conformance.sh:9]: If somebody wants to stop the process by force, why would we gracefully cleanup?

File: hack/conformance.sh [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+echo "Extracting the full report and keep it under ${ARTIFACTS}"
+tar xf ${results_archive} -C ${ARTIFACTS}
+cp ${ARTIFACTS}/plugins/kubevirt-conformance/results/global/junit.xml ${ARTIFACTS}/junit.conformance.xml
+
+echo 'Evaluating success of the test run'
+sonobuoy results ${results_archive} | grep -q 'Status: passed' || {
+    echo 'Conformance suite has failed'
+    exit 1
+}
```
Comment (oshoval) [inline hack/conformance.sh:33]: maybe worth to add a "success" otherwise?
or just retcode 0 and not errors printed would be enough

File: hack/dockerized [hunk 5]
```diff
@@ -22,7 +22,9 @@ if [ -z ${KUBEVIRT_CRI} ]; then
     fi
 fi
 
-KUBEVIRT_BUILDER_IMAGE="kubevirt/builder@sha256:1f521ae568f4bdbcb41f02f9f11a7b5234120cf7c6a673df4f42b6fdfe4c35d3"
+# TODO
```
Comment (oshoval) [inline hack/dockerized:25]: TODO and the next line should stay?

File: manifests/release/conformance.yaml.in [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+spec:
+  command:
+  - /usr/bin/conformance
+  image: {{.DockerPrefix}}/conformance:{{.DockerTag}}
+  name: plugin
+  resources: {}
+  volumeMounts:
+  - mountPath: /tmp/results
+    name: results
+
```
Comment (oshoval) [inline manifests/release/conformance.yaml.in:14]: do we need this line ?

File: hack/dockerized (hunk 5 above)
Comment (phoracek) [reply hack/dockerized:25]: See the description of the PR

File: manifests/release/conformance.yaml.in (hunk 6 above)
Comment (phoracek) [reply manifests/release/conformance.yaml.in:14]: Dropped

File: automation/conformance.sh (hunk 3 above)
Comment (oshoval) [reply automation/conformance.sh:9]: both SIGINT and SIGQUIT are catchable, same reason you added SIGINT imo
the cleanup should be done (unless you consider it unhappy as the text said)

just in case he SIGKILL we cant and shouldnt clean

File: hack/conformance.sh (hunk 4 above)
Comment (phoracek) [reply hack/conformance.sh:33]: Done

**Comment by oshoval** [comment]
/lgtm
Thanks
(i understand there will be more pushes)

File: Makefile [hunk 7]
```diff
@@ -154,6 +157,7 @@ bump-kubevirtci:
 
 .PHONY: \
 	build-verify \
+	conformance \
```
Comment (oshoval) [inline Makefile:160]: worth to keep order imo ?

Comment (phoracek) [reply Makefile:160]: There is no order AFAIK

Comment (oshoval) [reply Makefile:160]: you right, just good practice to order it imo, not need to be in this PR, 
just if you want, to put it before `clean` as the target it self

Comment (phoracek) [reply Makefile:160]: I see no reason to try to keep an order in an unsorted list.

**Comment by phoracek** [comment]
/hold cancel

The image was published

File: hack/conformance.sh [hunk 8]
```diff
@@ -0,0 +1,34 @@
+#!/usr/bin/env bash
+#
+# This script should be executed through the makefile via `make conformance`.
+
+set -euo pipefail
```
Comment (RamLavi) [inline hack/conformance.sh:5]: why don't we use the x flag?

File: manifests/release/conformance.yaml.in [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+  result-format: junit
+spec:
+  command:
+  - /usr/bin/conformance
+  image: {{.DockerPrefix}}/conformance:{{.DockerTag}}
+  name: plugin
+  resources: {}
+  volumeMounts:
+  - mountPath: /tmp/results
+    name: results
```
Comment (RamLavi) [inline manifests/release/conformance.yaml.in:13]: running 
```
sonobuoy gen plugin --name kubevirt-conformance --cmd /usr/bin/conformance --image IMAGE -f junit > manifests/release/conformance.yaml.in
sed -i 's#IMAGE#{{.DockerPrefix}}/conformance:{{.DockerTag}}#' manifests/release/conformance.yaml.in
```

gives another `\n` to this file from some reason. could this be an issue (prorclain/check wise)?

Comment (phoracek) [reply manifests/release/conformance.yaml.in:13]: It is not autogenerated, so I don't think it would cause any harm. I can keep the line if you think it would be better, I don't have a strong feelings about it.

Comment (RamLavi) [reply manifests/release/conformance.yaml.in:13]: if you don't think it could cause an issue then I prefer what you did. 
we can always change it later should a problem arises.

File: hack/conformance.sh (hunk 8 above)
Comment (phoracek) [reply hack/conformance.sh:5]: I wanted to make it pretty by showing `echo` logs when running locally. Would you find it more useful if it had `-x`? I don't mind adding it.

Comment (RamLavi) [reply hack/conformance.sh:5]: I get what you mean, but I think in the end these prints can be useful to see what actually happens. so I would vote to add it.

File: api/openapi-spec/swagger.json [hunk 10]
```diff
@@ -8106,8 +8106,21 @@
...
+     "resourceName": {
+      "type": "string"
+     },
+     "resourceType": {
+      "type": "string"
+     },
+     "type": {
+      "type": "string"
+     }
+    }
```
Comment (qinqon) [inline api/openapi-spec/swagger.json:8123]: Is this related ?

**Review by qinqon — CHANGES_REQUESTED** [review]
Small nits

File: automation/conformance.sh [hunk 11]
```diff
@@ -0,0 +1,13 @@
+#!/usr/bin/env bash
+
+set -euo pipefail
+
+export WORKSPACE="${WORKSPACE:-$PWD}"
+readonly ARTIFACTS_PATH="${ARTIFACTS-$WORKSPACE/exported-artifacts}"
+mkdir -p "$ARTIFACTS_PATH"
+
+trap "{ make cluster-down; cp -r _out/artifacts/conformance ${ARTIFACTS_PATH}; }" EXIT SIGINT SIGTERM SIGQUIT
```
[the remark this answers is not in the document]
Comment (oshoval) [reply automation/conformance.sh:9]: depends if the file has a shebang that runs it with bash, else on dash (on prow) it wont work according some link we saw.
btw SIGINT is needed, as was tried even on bash.

[the remark this answers is not in the document]
Comment (phoracek) [reply automation/conformance.sh:9]: https://github.com/kubevirt/kubevirt/pull/4115#discussion_r484034330

File: api/openapi-spec/swagger.json (hunk 10 above)
Comment (phoracek) [reply api/openapi-spec/swagger.json:8123]: Nope, dropping

File: hack/builder/Dockerfile (hunk 1 above)
[the remark this answers is not in the document]
Comment (phoracek) [reply hack/builder/Dockerfile:97]: Done

File: hack/conformance.sh (hunk 8 above)
Comment (phoracek) [reply hack/conformance.sh:5]: Added

File: hack/conformance.sh [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+echo 'Executing conformance tests and wait for them to finish'
+sonobuoy run --wait --plugin _out/manifests/release/conformance.yaml
+
+trap "{ echo 'Cleaning up after the test execution'; sonobuoy delete --wait; }" EXIT SIGINT SIGTERM SIGQUIT
+
+echo 'Downloading report about the test execution'
+results_archive=${ARTIFACTS}/$(cd ${ARTIFACTS} && sonobuoy retrieve)
+
+echo 'Results:'
+sonobuoy results ${results_archive}
```
[the remark this answers is not in the document]
Comment (phoracek) [reply hack/conformance.sh:23]: They are a part of artifacts, is that enough?

File: api/openapi-spec/swagger.json (hunk 10 above)
Comment (oshoval) [reply api/openapi-spec/swagger.json]: this is the flaky generate causing it, dont try to fight it, its WIP

**Comment by oshoval** [comment]
/lgtm
from me, thanks

File: automation/conformance.sh (hunk 11 above)
[the remark this answers is not in the document]
Comment (qinqon) [reply automation/conformance.sh:9]: ack

File: hack/conformance.sh (hunk 12 above)
[the remark this answers is not in the document]
Comment (qinqon) [reply hack/conformance.sh:23]: Then is all good I suppose results is dumping logs too right ?

[the remark this answers is not in the document]
Comment (phoracek) [reply hack/conformance.sh:23]: They do

File: automation/conformance.sh [hunk 13]
```diff
@@ -0,0 +1,13 @@
+#!/usr/bin/env bash
+
+set -exuo pipefail
```
Comment (AlonaKaplan) [inline automation/conformance.sh:3]: I'm not a big bash expert, but it is written [here](https://vaneyckt.io/posts/safer_bash_scripts_with_set_euxo_pipefail/) that -E is required when you inherit a trap, so that the -e won't cause the script to exit before the trap is fired.

File: hack/conformance.sh [hunk 14]
```diff
@@ -0,0 +1,34 @@
+#!/usr/bin/env bash
+#
+# This script should be executed through the makefile via `make conformance`.
+
+set -exuo pipefail
```
Comment (AlonaKaplan) [inline hack/conformance.sh:5]: same regarding -E

File: automation/conformance.sh (hunk 13 above)
Comment (oshoval) [reply automation/conformance.sh:3]: just if running in a subshell, else its safe

Comment (phoracek) [reply automation/conformance.sh]: It also applies to called shell functions. We may not need it in the current state of our scripts but it is a safe option to add it.

Done

File: hack/conformance.sh (hunk 14 above)
Comment (phoracek) [reply hack/conformance.sh]: Done

File: automation/conformance.sh (hunk 13 above)
Comment (oshoval) [reply automation/conformance.sh]: fyi, on other PRs it was concluded that it does work on shell functions (we even binded to bash here), when i raised it (with the same url even).
up to you, i just think we should line up on a method.

Comment (phoracek) [reply automation/conformance.sh]: Ok, then let's keep it in case we ever introduce a subshell to the script. Was that tested by you too?

Comment (oshoval) [reply automation/conformance.sh]: not in that PR mentioned scope, it was tested ad hoc by the author.

Comment (phoracek) [reply automation/conformance.sh]: subshell of shell function?

Comment (oshoval) [reply automation/conformance.sh]: he tested both:
* sub shell - needs -E
* shell function (he used bash from what i remember) - doesnt need -E

Comment (phoracek) [reply automation/conformance.sh]: Who is "he", do you have a reference?

Comment (oshoval) [reply automation/conformance.sh]: if i find the threads i will paste

Comment (oshoval) [reply automation/conformance.sh]: shell functions
kubevirt/kubevirtci#428 (comment)

subshell
kubevirt/kubevirtci#428 (comment)

Comment (phoracek) [reply automation/conformance.sh]: You linked the PR twice, could you give me link to the specific comment?

Comment (oshoval) [reply automation/conformance.sh]: copy paste removed the link
shell functions (whole discussion, bottom line in the end)
https://github.com/kubevirt/kubevirtci/pull/428#discussion_r476229164

subshell (bottom line)
https://github.com/kubevirt/kubevirtci/pull/428#discussion_r476558135

Comment (phoracek) [reply automation/conformance.sh]: In the second link, @ormergi commented that -E **is** needed. Am I reading it wrong?

Comment (oshoval) [reply automation/conformance.sh]: in subshell only, see first that say its not needed in function (see bottom of first thread please)
exactly as i said above.

Comment (phoracek) [reply automation/conformance.sh]: So I want to keep -E in case anybody ever introduces a subshell into one of these scripts.

Sounds good to you?

Comment (oshoval) [reply automation/conformance.sh]: yes thx, with the only opinion that we should stick to it in my opinion on every PR we are creating, 
its easier, safer and saves time
(i understand some might think otherwise)

else i dont see why to put stuff for future, unless its per Author decision, which might waste time on future discussion.

up to you, thx

Comment (phoracek) [reply automation/conformance.sh]: I think of this as a valid reset to make Bash work in an expected manner, no matter if all the code in the scripts needs it or not.

If you are ok with this shape, consider granting lgtm.

**Review by oshoval — APPROVED** [review]
